### ISSUE TITLE
Image normalization in Yolov8?

### ISSUE THREAD

**Issue by tjasmin111** [ultralytics/ultralytics#6979]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

What is Yolov8 (nano, detection) image normalization approach? Are image data normalized to [0,1] like common approaches or maybe [-0.5, +0.5]?

**Comment by glenn-jocher** [comment]
@tjasmin111 hello! In YOLOv8, we normalize image data to the [0, 1] range. This normalization step is quite standard and helps with training stability and performance. If you have any more questions or need further clarification, feel free to check the Ultralytics Docs or ask us here. Happy detecting! 😊

**Comment by liuqinglong110** [comment]
> @tjasmin111 hello! In YOLOv8, we normalize image data to the [0, 1] range. This normalization step is quite standard and helps with training stability and performance. If you have any more questions or need further clarification, feel free to check the Ultralytics Docs or ask us here. Happy detecting! 😊

Hello, I am deploying the model generated by training to mobile hardware. In this process, I want to know the preprocessing method or normalization method used by default in the ultralytics (yolov8n-seg) code. Therefore, only if the training process and prediction process use the same normalization or preprocessing method can the model function better.
In the past, when I used other algorithms for training, because the training framework was not smart enough, I needed to calculate the mean and std of the training set myself before starting training. But during the training process of ultralytics (yolov8n-seg), I did not find the location where mean and std need to be set. It seems that there is no need to set mean and std.
I searched the ultralytics (yolov8n-seg) code, trying to find the specific values of mean and std during training. I'm really sorry, I didn't find what I was looking for. I saw that the default values of mean and std in some functions are mean=(0.0, 0.0, 0.0) and std=(1.0, 1.0, 1.0) respectively. 
During the actual training process, I use opencv to read RGB images. Does this mean that during the ultralytics (yolov8n-seg) training process, the code is just normalized by dividing by 255? Since I saw some of your answers to other questions todayhttps://github.com/ultralytics/ultralytics/issues/2819, I found that there might be a problem with my previous understanding, so I wanted to confirm with you.

I saw some of your suggestions in other questions, like the one pictured below:https://github.com/ultralytics/ultralytics/issues/2819
![image](https://github.com/ultralytics/ultralytics/assets/23652342/c2517097-e406-4a8e-8f66-49eafba93c25)
Therefore, I want to know how these two values ​​are calculated during the training process of ultralytics (yolov8n-seg) when I do not actively set or modify mean and std.
Looking forward to your answer, thanks again！

**Comment by JinyuJinyuJinyu** [comment]
> > @tjasmin111 hello! In YOLOv8, we normalize image data to the [0, 1] range. This normalization step is quite standard and helps with training stability and performance. If you have any more questions or need further clarification, feel free to check the Ultralytics Docs or ask us here. Happy detecting! 😊
> 
> Hello, I am deploying the model generated by training to mobile hardware. In this process, I want to know the preprocessing method or normalization method used by default in the ultralytics (yolov8n-seg) code. Therefore, only if the training process and prediction process use the same normalization or preprocessing method can the model function better. In the past, when I used other algorithms for training, because the training framework was not smart enough, I needed to calculate the mean and std of the training set myself before starting training. But during the training process of ultralytics (yolov8n-seg), I did not find the location where mean and std need to be set. It seems that there is no need to set mean and std. I searched the ultralytics (yolov8n-seg) code, trying to find the specific values of mean and std during training. I'm really sorry, I didn't find what I was looking for. I saw that the default values of mean and std in some functions are mean=(0.0, 0.0, 0.0) and std=(1.0, 1.0, 1.0) respectively. During the actual training process, I use opencv to read RGB images. Does this mean that during the ultralytics (yolov8n-seg) training process, the code is just normalized by dividing by 255? Since I saw some of your answers to other questions todayhttps://github.com/[/issues/2819](https://github.com/ultralytics/ultralytics/issues/2819), I found that there might be a problem with my previous understanding, so I wanted to confirm with you.
> 
> I saw some of your suggestions in other questions, like the one pictured below:#2819 ![image](https://private-user-images.githubusercontent.com/23652342/293868234-c2517097-e406-4a8e-8f66-49eafba93c25.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM0MDk0MDEsIm5iZiI6MTcxMzQwOTEwMSwicGF0aCI6Ii8yMzY1MjM0Mi8yOTM4NjgyMzQtYzI1MTcwOTctZTQwNi00YThlLThmNjYtNDllYWZiYTkzYzI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE4VDAyNTgyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyZDhhZWU0ZjhiNzVkMjNkNjkwNjJmYTczNWY1OGYzNjI4YTlkZTM1NmU2ODBkN2Y4ZTIzODVhZGE2OGQ4OWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4n-TFcFe26jt393fDMgxK8XRQv6pPMCxtlMd3HQzkNk) Therefore, I want to know how these two values ​​are calculated during the training process of ultralytics (yolov8n-seg) when I do not actively set or modify mean and std. Looking forward to your answer, thanks again！

The normalization is done by dividing 255. if u want to implement z-score normalization (mean and std), u have to modify the code or write on a separate .py file.

**Comment by SengeNB** [comment]
If I want to normalize my dataset using the Z-score method, does YOLOv8 directly support this functionality? Thank you.

**Comment by glenn-jocher** [comment]
YOLOv8 does not directly support Z-score normalization. You'll need to implement it separately in your preprocessing pipeline.

**Comment by srisaisubramanyam** [comment]
@glenn-jocher
Suppose I want to normalize each image individually before sending it for training in every batch. Could you please suggest the best way to implement this?
Thanks

**Comment by glenn-jocher** [comment]
@srisaisubramanyam Easiest is to override the trainer’s `preprocess_batch()` and apply per-image (per-channel) z-score after the default scaling step; see the function in the detection trainer reference and the loader’s `ToTensor` behavior for context: [preprocess_batch](https://docs.ultralytics.com/reference/models/yolo/detect/train/#preprocess_batch), [ToTensor](https://docs.ultralytics.com/reference/data/augment/#ultralytics.data.augment.ToTensor).

```python
# my_zscore_trainer.py
import torch
from ultralytics.models.yolo.detect.train import DetectionTrainer

class ZScoreTrainer(DetectionTrainer):
    def preprocess_batch(self, batch: dict) -> dict:
        batch = super().preprocess_batch(batch)  # moves to device + scales
        imgs = batch["img"]  # (B, C, H, W) in [0,1]
        mean = imgs.mean(dim=(2, 3), keepdim=True)
        std = imgs.std(dim=(2, 3), keepdim=True).clamp_min(1e-6)
        batch["img"] = (imgs - mean) / std
        return batch
```

```python
# train.py
from ultralytics import YOLO
from my_zscore_trainer import ZScoreTrainer

model = YOLO("yolov8n-seg.pt")
model.train(data="data.yaml", imgsz=640, trainer=ZScoreTrainer)
```

Be sure to apply the same normalization in your mobile inference pipeline to match training.
